tests/lib/rules/no-unnecessary-act.test.ts

```diff
@@ -0,0 +1,782 @@
+  valid: [
+    {
+      code: `// case: RTL act wrapping non-RTL calls
+      import { act } from '@testing-library/react'
+
+      test('valid case', async () => {
+        act(() => {
+          stuffThatDoesNotUseRTL();
+        });
+
+        await act(async () => {
+          stuffThatDoesNotUseRTL();
+        });
+
+        act(function() {
+          stuffThatDoesNotUseRTL();
+          const a = foo();
+        });
+
+        act(function() {
+          return stuffThatDoesNotUseRTL();
+        });
+
+        act(() => stuffThatDoesNotUseRTL());
+      });
+      `,
+    },
+    {
+      code: `// case: RTU act wrapping non-RTL
+      import { act } from 'react-dom/test-utils'
+
+      test('valid case', async () => {
+        act(() => {
+          stuffThatDoesNotUseRTL();
+        });
+
+        await act(async () => {
+          stuffThatDoesNotUseRTL();
+        });
+
+        act(function() {
+          stuffThatDoesNotUseRTL();
+        });
+
+        act(function() {
+          return stuffThatDoesNotUseRTL();
+        });
+
+        act(() => stuffThatDoesNotUseRTL());
+      });
+      `,
+    },
+    {
+      settings: {
+        'testing-library/utils-module': 'test-utils',
+      },
+      code: `// case: RTL act wrapping non-RTL - AGR disabled
+      import { act } from '@testing-library/react'
+      import { waitFor } from 'somewhere-else'
+
+      test('valid case', async () => {
+        act(() => {
+          waitFor();
+        });
+
+        await act(async () => {
+          waitFor();
+        });
+
+        act(function() {
+          waitFor();
+        });
+
+        act(function() {
+          return waitFor();
+        });
+
+        act(() => waitFor());
+      });
+      `,
+    },
+    {
+      code: `// case: RTL act wrapping both RTL and non-RTL calls
+      import { act, render, waitFor } from '@testing-library/react'
+
+      test('valid case', async () => {
+        act(() => {
+          render(element);
+          stuffThatDoesNotUseRTL();
+        });
+
+        await act(async () => {
+          waitFor();
+          stuffThatDoesNotUseRTL();
+        });
+
+        act(function() {
+          waitFor();
+          stuffThatDoesNotUseRTL();
+        });
+      });
+      `,
+    },
+    {
+      settings: {
+        'testing-library/utils-module': 'test-utils',
+      },
+      code: `// case: non-RTL act wrapping RTL - AGR disabled
+      import { act } from 'somewhere-else'
+      import { waitFor } from '@testing-library/react'
+
+      test('valid case', async () => {
+        act(() => {
+          waitFor();
+        });
+
+        await act(async () => {
+          waitFor();
+        });
+
+        act(function() {
+          waitFor();
+        });
+
+        act(function() {
+          return waitFor();
+        });
+
+        act(() => waitFor());
+
+        act(() => {})
+        await act(async () => {})
+        act(function() {})
+      });
+      `,
+    },
+  ],
```

Comment:
in my current project, we are using promises instead of async/await, maybe it's worth adding a case that considers using `.then()` instead of `await` to add more confidence

Reply:
Sure. Just out of curiosity: are you chaining several `.then()` in those tests? If so, doesn't that make the code really messy?

Reply:
I just joined the project and the team prefer a more functional approach chaining `.then()` as they think async/await looks more imperative. If well chained, it does not look messy 😛

Reply:
Not arguing against it, but I can't imagine writing tests chaining `.then()`! Anyway, please share any feedback you have if find anything to improve around thenables promises.

Reply:
I will! the project is fairly new (startup mode 😅 ) so we are not using RTL yet, but when the time comes, this plugin will be added and feedback be gathered 💪